src/Plugin/Commerce/PaymentGateway/ExpressCheckout.php

```diff
@@ -0,0 +1,570 @@
+      $container->get('plugin.manager.commerce_payment_type'),
+      $container->get('plugin.manager.commerce_payment_method_type'),
+      $container->get('http_client'),
+      $container->get('state')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration() {
+    return [
+      'api_username' => '',
+      'api_password' => '',
+      'signature' => '',
+      'solution_type' => 'Mark',
+    ] + parent::defaultConfiguration();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildConfigurationForm($form, $form_state);
+
+    $form['api_username'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('API Username'),
+      '#default_value' => $this->configuration['api_username'],
+      '#required' => TRUE,
+    ];
+    $form['api_password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('API Password'),
+      '#default_value' => $this->configuration['api_password'],
+      '#required' => TRUE,
+    ];
+    $form['signature'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Signature'),
+      '#default_value' => $this->configuration['signature'],
+      '#required' => TRUE,
+    ];
+
+    $form['solution_type'] = [
+      '#type' => 'radios',
+      '#title' => t('Type of checkout flow'),
+      '#description' => t('Express Checkout Account Optional (ECAO) where PayPal accounts are not required for payment may not be available in all markets.'),
+      '#options' => [
+        'Mark' => t('Require a PayPal account (this is the standard configuration).'),
+        'SoleLogin' => t('Allow PayPal AND credit card payments, defaulting to the PayPal form.'),
+        'SoleBilling' => t('Allow PayPal AND credit card payments, defaulting to the credit card form.'),
+      ],
+      '#default_value' => $this->configuration['solution_type'],
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    parent::submitConfigurationForm($form, $form_state);
+    if (!$form_state->getErrors()) {
+      $values = $form_state->getValue($form['#parents']);
+      $this->configuration['api_username'] = $values['api_username'];
+      $this->configuration['api_password'] = $values['api_password'];
+      $this->configuration['signature'] = $values['signature'];
+      $this->configuration['solution_type'] = $values['solution_type'];
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function onReturn(OrderInterface $order, Request $request) {
+    // GetExpressCheckoutDetails API Operation (NVP).
+    // Shows information about an Express Checkout transaction.
+    $paypal_response = $this->apiGetExpressCheckoutDetails($order);
+
+    // If the request failed, exit now with a failure message.
+    if ($paypal_response['ACK'] == 'Failure') {
+      return FALSE;
```

Comment:
Per the `OffsitePaymentGatewayInterface` documentation:

```
   * @throws \Drupal\commerce_payment\Exception\PaymentGatewayException
   *   Thrown when the request is invalid or the payment failed.
```

We need to throw an exception and not just return `FALSE`.